include/ginkgo/core/preconditioner/isai.hpp

```diff
@@ -0,0 +1,207 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_PRECONDITIONER_ISAI_HPP_
+#define GKO_CORE_PRECONDITIONER_ISAI_HPP_
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/composition.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+/**
+ * @brief The Preconditioner namespace.
+ *
+ * @ingroup precond
+ */
+namespace preconditioner {
+
+
+/**
+ * The Incomplete Sparse Approximate Inverse (ISAI) Preconditioner generates
+ * approximate inverse matrices for a given lower triangular matrix L and upper
+ * triangular matrix U. Using the precionditioner computes $aiU * aiL * x$
+ * for a given vector x (may have multiple right hand sides). aiU and aiL
+ * are the approximate inverses for U and L respectively.
+ *
+ * The sparsity pattern used for the approximate inverses is the same as
+ * the sparsity pattern of the respective triangular matrix.
+ * The L and U matrices need to be in a Composition<ValueType> object in the
+ * order: L, U.
+ *
+ * For more details on the algorithm, see the paper
+ * <a href="https://doi.org/10.1016/j.parco.2017.10.003">
+ * Incomplete Sparse Approximate Inverses for Parallel Preconditioning</a>,
+ * which is the basis for this work.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ * @ingroup isai
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Isai : public EnableLinOp<Isai<ValueType, IndexType>> {
```

Comment:
This class combines two (imho completely distinct) responsibilities: The ISAI itself and the functionality of a `Composition`. I would argue it is cleaner to keep them separate and give the ISAI another factory parameter (or figure out a more general way of attaching *type* flags to matrices like with CUSPARSE_FILL_UPPER) to determine if it's a lower or upper triangular matrix.

Reply:
I think matrix tags in general would be a good idea to add to `LinOp` objects. I added it to the Ginkgo 2.0 Issue #402 
I will also add the possibility to generate the Preconditioner for only a lower or an upper triangular matrix.

Reply:
I have moved this to #495